### ISSUE TITLE
`_onprogress` raises a protocol error for a race servers cannot avoid (progress notification arriving with the response)

### ISSUE THREAD

**Issue by sweetrb** [modelcontextprotocol/typescript-sdk#2580]
### Summary

When a request completes, `Protocol._onresponse` deletes that request's progress handler. Any progress notification for the token that is processed afterwards falls into `_onprogress`, finds no handler, and is reported through `_onerror`:

```
Received a progress notification for an unknown token: {...}
```

Discarding the notification is reasonable — once the request is done, progress for it is moot. **Raising a protocol error for it is not**, because a server has no way to avoid producing that condition.

### Why a server cannot avoid it

A server that reports progress for a batch operation naturally emits a terminal `progress === total` notification just before returning the result. Those two writes are adjacent by construction. Whether the client processes the notification before or after the response is a timing race the server does not control — and on a loaded machine it loses.

Concretely, in `src/shared/protocol.ts`:

```ts
// _onresponse
if (!isTaskResponse) {
  this._progressHandlers.delete(messageId);
}

// _onprogress
const handler = this._progressHandlers.get(messageId);
if (!handler) {
  this._onerror(new Error(`Received a progress notification for an unknown token: ...`));
  return;
}
```

So any server emitting a terminal progress notification will intermittently cause `onerror` to fire in every client. For hosts that log or surface protocol errors, a well-behaved server ends up manufacturing spurious error noise on a successful operation.

Attempting to fix it server-side does not work. Awaiting the notification sends before returning the result orders the *writes* correctly, but cannot prevent the client from tearing down its handler before it dispatches them.

### Reproduction

Server emits N progress notifications then the result, with no artificial spacing between them. Client:

```ts
const progress: unknown[] = [];
const errors: string[] = [];
client.onerror = (e) => errors.push(e.message);

await client.callTool({ name: "export", arguments: { /* 3 items */ } },
  CallToolResultSchema, { onprogress: (p) => progress.push(p) });
```

Observed (server sent **4** notifications, all before the result):

```
received=1
errors=[
  'Received a progress notification for an unknown token: {"method":"notifications/progress","params":{"progress":1,"total":3,...,"progressToken":1}}',
  '...{"progress":2,...}',
  '...{"progress":3,...}'
]
```

One delivered, three discarded with an error each. With ~100ms spacing between notifications the same code delivers all four — i.e. it is purely a timing race, not a protocol violation by either side.

Observed with `@modelcontextprotocol/sdk` 1.29.0 over stdio.

### Suggested change

Treat "progress notification for a request that has just completed" as an expected, benign race rather than an error:

- don't route it to `_onerror`; ignore it silently, or log at debug level; **or**
- keep a short grace period after response for recently-completed tokens and drop matching notifications quietly.

Either removes the false error without changing the (sensible) decision not to deliver late progress.

### Secondary question

Is there any supported way for a server to deliver a terminal progress notification reliably? If not, that is worth stating in the spec/docs, so server authors know the final `progress === total` frame is inherently best-effort and clients know not to build completion UI on it. Right now it looks reliable, and fails only under load.

**Comment by qiushui7** [comment]
Hi! I’d like to work on this issue.

My proposed approach is to maintain a bounded cache of recently completed progress tokens. Late progress notifications associated with those tokens would be ignored without invoking `onerror`, while notifications with genuinely unknown tokens would continue to trigger the existing protocol error behavior.

I’d also add a deterministic regression test covering progress notifications that arrive immediately before or after the corresponding response, along with a test confirming that unrelated unknown tokens still report an error.

Would this approach be acceptable?

**Comment by sweetrb** [comment]
Yes — that's the right shape, and thanks for picking it up.

Keeping the error for genuinely unknown tokens is the important half. The current message is a useful signal when a server really does send garbage; the problem is only that a token which was valid moments ago is indistinguishable from one that never existed. A recently-completed set separates those two cases cleanly.

Two suggestions:

**The regression test can be fully deterministic — don't try to reproduce the race.** My repro needed a loaded CI runner because I was racing a real server against a real client, but the behaviour under test isn't actually timing-dependent: it's just "what happens when `_onprogress` runs after `_onresponse` for the same id". Driving those two in that order against a `Protocol` instance with an in-memory transport pins it exactly, with no sleeps and no flakiness. Worth covering three cases:

- progress **before** response → delivered to `onprogress` (guards against over-eager eviction)
- progress **after** response, same id → silently ignored, `onerror` **not** called
- progress for an id that was never issued → still reports the error

**Worth stating the eviction policy explicitly.** A long-lived client can issue a lot of requests, so the cache needs a bound that's obvious from the code — a small fixed-size FIFO/LRU, or a short time window, either is fine. Only two details actually matter: it must not grow without limit, and the retention window should comfortably exceed the transport's reorder/dispatch gap, which is sub-millisecond in practice. Even a very small cache (say the last few dozen ids) would have covered every case I saw.

One thing you may want to decide deliberately: whether an ignored late notification is *completely* silent or logged at debug level. I'd lean debug-loggable rather than silent, so the "am I losing progress notifications?" question stays answerable without a patched SDK — that's precisely the question I couldn't answer without instrumenting a client by hand.

Happy to test a branch against the server that surfaced this (a stdio MCP server emitting a terminal `progress === total` immediately before its result), if a real-world check is useful alongside the unit tests.

**Comment by qiushui7** [comment]
Thanks — this is very helpful.

I’ll use deterministic in-memory transport tests for the three cases:

1. A progress notification for the request’s registered progress token before the response is delivered normally.

2. A progress notification after the response, using the completed request’s progress token, is ignored without calling `onerror`.

3. A progress notification with a token that was never registered still reports the existing unknown-token error.

For eviction, I’m leaning toward a small fixed-size FIFO of recently completed progress tokens, probably 64 entries. That keeps the bound explicit, requires no timers, and should comfortably cover the transport dispatch gap.

I’ll also check whether the SDK already has an appropriate internal debug hook. If it does, I’ll use it for ignored late notifications. Otherwise, I’ll keep that path silent and document the intentional drop in code rather than introducing console logging from the library.

Since the reproduction is against 1.29.0 and the relevant implementation is on the `v1.x` branch, I’ll target `v1.x` unless the maintainers prefer otherwise.

I’ll share the branch once the tests and implementation are ready. Thanks for offering to validate it against the real stdio server.

**Comment by qiushui7** [comment]
The implementation is now available in draft PR #2586, and all official CI checks are passing.

A preview package is available for real-world testing:

```bash
npm i https://pkg.pr.new/@modelcontextprotocol/sdk@2586
```

@sweetrb, if you’re still available, could you test it against the stdio server that originally surfaced the race? In particular, it would be helpful to confirm that the late terminal progress notifications no longer trigger `onerror`, while the request still completes normally.

Once that validation is complete, I’ll mark the PR ready for review.
